src/runtime/hexagon_remote/host_malloc.cpp

```diff
@@ -40,21 +51,43 @@ struct ion_handle_data {
 };
 
 #define ION_IOC_ALLOC _IOWR('I', 0, ion_allocation_data)
+#define ION_IOC_ALLOC_NEWER _IOWR('I', 0, ion_allocation_data_newer)
 #define ION_IOC_FREE _IOWR('I', 1, ion_handle_data)
 #define ION_IOC_MAP _IOWR('I', 2, ion_fd_data)
 
+// ION IOCTL approach
+// This function will first try older IOCTL approach provided we have not determined
+// that we should use the newer IOCTL. Once we have determined we should use the
+// newer IOCTL method , we cease calling the older IOCTL.
 ion_user_handle_t ion_alloc(int ion_fd, size_t len, size_t align, unsigned int heap_id_mask, unsigned int flags) {
-    ion_allocation_data alloc = {
+
+    if (!use_newer_ioctl) {
+        ion_allocation_data alloc = {
+            len,
+            align,
+            heap_id_mask,
+            flags,
+            0
+        };
+        if (ioctl(ion_fd, ION_IOC_ALLOC, &alloc) >= 0) {
+            return alloc.handle;
+        }
+    }
```

Comment:
Of course, we can try the old ioctl, and then fallback to new one if it fails.
This will mean to IOCTL calls for every ion_alloc for newer devices going forward.
if this is deemed low cost, then sure i will make the change which simplifies all the code.

Reply:
I would try measuring it. If it's expensive, you can still "cache" the result in a global (or static local, since this doesn't need to span multiple functions anymore) state bool, which is still a lot simpler than what is there now.

Reply:
btw. there are calls to ion_free which should not be made if we are using the new ion ioctl approach. These result in failures which show up in logcat output. i can either guard these calls to ion_free with the use_newer_ioctl, or i can suppress the messages and simply make the calls. I would prefer the former approach just in case the old ioctl actually did encounter an error calling ion_free.

Reply:
hmm, I missed that. I'd keep the use_newer_ioctl flag, but I would still change the logic to switch to using it as I suggested (so you don't need the dummy malloc).

Reply:
yes, i like your simplified approach, i will definitely do it that way.